File: apps/electron/scripts/generate-installer-assets.mjs
#!/usr/bin/env node

/**
 * Generate installer assets for macOS DMG and Windows NSIS
 *
 * Creates:
 * - DMG background with "drag to Applications" design
 * - NSIS installer header and sidebar images
 */

import { execSync } from "child_process"
import { existsSync, mkdirSync, writeFileSync } from "fs"
import { join, dirname } from "path"
import { fileURLToPath } from "url"

const __filename = fileURLToPath(import.meta.url)
const __dirname = dirname(__filename)
const buildDir = join(__dirname, "../build")

// Ensure build directory exists
if (!existsSync(buildDir)) {
  mkdirSync(buildDir, { recursive: true })
}

/**
 * Generate DMG background image using ImageMagick or sharp
 * Size: 660x400 pixels
 */
async function generateDMGBackground() {
  const outputPath = join(buildDir, "dmg-background.png")

  // Create SVG template
  const svg = `<?xml version="1.0" encoding="UTF-8"?>
<svg width="660" height="400" xmlns="http://www.w3.org/2000/svg">
  <defs>
    <linearGradient id="bg" x1="0%" y1="0%" x2="0%" y2="100%">
      <stop offset="0%" style="stop-color:#1a1a2e;stop-opacity:1" />
      <stop offset="100%" style="stop-color:#16213e;stop-opacity:1" />
    </linearGradient>
    <filter id="shadow" x="-20%" y="-20%" width="140%" height="140%">
      <feDropShadow dx="0" dy="4" stdDeviation="8" flood-color="#000" flood-opacity="0.3"/>
    </filter>
  </defs>

  <!-- Background -->
  <rect width="660" height="400" fill="url(#bg)"/>

  <!-- Subtle grid pattern -->
  <pattern id="grid" width="30" height="30" patternUnits="userSpaceOnUse">
    <path d="M 30 0 L 0 0 0 30" fill="none" stroke="rgba(255,255,255,0.03)" stroke-width="1"/>
  </pattern>
  <rect width="660" height="400" fill="url(#grid)"/>

  <!-- App name at top -->
  <text x="330" y="55" font-family="SF Pro Display, -apple-system, BlinkMacSystemFont, sans-serif"
        font-size="28" font-weight="600" fill="#ffffff" text-anchor="middle">
    S-AGI
  </text>
  <text x="330" y="80" font-family="SF Pro Text, -apple-system, BlinkMacSystemFont, sans-serif"
        font-size="13" fill="rgba(255,255,255,0.6)" text-anchor="middle">
    AI Agent for Spreadsheets
  </text>

  <!-- Arrow pointing from app to Applications -->
  <g transform="translate(330, 170)">
    <!-- Arrow line -->
    <line x1="-70" y1="0" x2="70" y2="0" stroke="rgba(255,255,255,0.4)" stroke-width="2" stroke-dasharray="8,4"/>
    <!-- Arrow head -->
    <polygon points="80,0 60,-12 60,12" fill="rgba(255,255,255,0.5)"/>
  </g>

  <!-- Drop zones indicators -->
  <g transform="translate(180, 170)" filter="url(#shadow)">
    <rect x="-50" y="-50" width="100" height="100" rx="20" fill="rgba(255,255,255,0.08)" stroke="rgba(255,255,255,0.15)" stroke-width="1"/>
  </g>
  <g transform="translate(480, 170)" filter="url(#shadow)">
    <rect x="-50" y="-50" width="100" height="100" rx="20" fill="rgba(255,255,255,0.08)" stroke="rgba(255,255,255,0.15)" stroke-width="1"/>
  </g>

  <!-- Labels -->
  <text x="180" y="260" font-family="SF Pro Text, -apple-system, BlinkMacSystemFont, sans-serif"
        font-size="12" fill="rgba(255,255,255,0.7)" text-anchor="middle">
    S-AGI.app
  </text>
  <text x="480" y="260" font-family="SF Pro Text, -apple-system, BlinkMacSystemFont, sans-serif"
        font-size="12" fill="rgba(255,255,255,0.7)" text-anchor="middle">
    Applications
  </text>

  <!-- Instructions -->
  <text x="330" y="340" font-family="SF Pro Text, -apple-system, BlinkMacSystemFont, sans-serif"
        font-size="14" fill="rgba(255,255,255,0.5)" text-anchor="middle">
    Drag S-AGI to your Applications folder to install
  </text>

  <!-- Version badge -->
  <rect x="280" y="360" width="100" height="24" rx="12" fill="rgba(255,255,255,0.1)"/>
  <text x="330" y="377" font-family="SF Pro Text, -apple-system, BlinkMacSystemFont, sans-serif"
        font-size="11" fill="rgba(255,255,255,0.5)" text-anchor="middle">
    v0.1.0
  </text>
</svg>`

  const svgPath = join(buildDir, "dmg-background.svg")
  writeFileSync(svgPath, svg)

  // Try to convert SVG to PNG using available tools
  try {
    // Try using sharp via Node
    const sharp = await import("sharp")
    await sharp.default(Buffer.from(svg))
      .png()
      .toFile(outputPath)
    console.log(`✓ Generated DMG background: ${outputPath}`)
    return true
  } catch (e) {
    // Try ImageMagick
    try {
      execSync(`convert ${svgPath} ${outputPath}`, { stdio: "pipe" })
      console.log(`✓ Generated DMG background (ImageMagick): ${outputPath}`)
      return true
    } catch (e2) {
      // Try rsvg-convert
      try {
        execSync(`rsvg-convert ${svgPath} -o ${outputPath}`, { stdio: "pipe" })
        console.log(`✓ Generated DMG background (rsvg): ${outputPath}`)
        return true
      } catch (e3) {
        console.log(`⚠ Could not convert SVG to PNG. SVG saved at: ${svgPath}`)
        console.log("  Install sharp, ImageMagick, or librsvg to auto-convert")
        return false
      }
    }
  }
}

/**
 * Generate NSIS installer header image
 * Size: 150x57 pixels (BMP format)
 */
async function generateNSISHeader() {
  const outputPath = join(buildDir, "installer-header.bmp")

  const svg = `<?xml version="1.0" encoding="UTF-8"?>
<svg width="150" height="57" xmlns="http://www.w3.org/2000/svg">
  <defs>
    <linearGradient id="headerBg" x1="0%" y1="0%" x2="100%" y2="0%">
      <stop offset="0%" style="stop-color:#1a1a2e;stop-opacity:1" />
      <stop offset="100%" style="stop-color:#0f3460;stop-opacity:1" />
    </linearGradient>
  </defs>
  <rect width="150" height="57" fill="url(#headerBg)"/>
  <text x="75" y="35" font-family="Segoe UI, Arial, sans-serif"
        font-size="18" font-weight="600" fill="#ffffff" text-anchor="middle">
    S-AGI
  </text>
</svg>`

  const svgPath = join(buildDir, "installer-header.svg")
  writeFileSync(svgPath, svg)

  try {
    const sharp = await import("sharp")
    // BMP requires raw pixel data, convert via PNG first
    const pngBuffer = await sharp.default(Buffer.from(svg))
      .resize(150, 57)
      .png()
      .toBuffer()

    // For BMP, we'll create a simple 24-bit BMP
    await sharp.default(pngBuffer)
      .toFormat('png')
      .toFile(outputPath.replace('.bmp', '.png'))

    // electron-builder can also use PNG for these
    console.log(`✓ Generated NSIS header: ${outputPath.replace('.bmp', '.png')}`)
    return true
  } catch (e) {
    console.log(`⚠ Could not generate NSIS header. SVG saved at: ${svgPath}`)
    return false
  }
}

/**
 * Generate NSIS installer sidebar image
 * Size: 164x314 pixels (BMP format)
 */
async function generateNSISSidebar() {
  const outputPath = join(buildDir, "installer-sidebar.bmp")

  const svg = `<?xml version="1.0" encoding="UTF-8"?>
<svg width="164" height="314" xmlns="http://www.w3.org/2000/svg">
  <defs>
    <linearGradient id="sidebarBg" x1="0%" y1="0%" x2="0%" y2="100%">
      <stop offset="0%" style="stop-color:#1a1a2e;stop-opacity:1" />
      <stop offset="50%" style="stop-color:#16213e;stop-opacity:1" />
      <stop offset="100%" style="stop-color:#0f3460;stop-opacity:1" />
    </linearGradient>
  </defs>
  <rect width="164" height="314" fill="url(#sidebarBg)"/>

  <!-- Logo circle -->
  <circle cx="82" cy="100" r="45" fill="rgba(255,255,255,0.1)" stroke="rgba(255,255,255,0.2)" stroke-width="2"/>
  <text x="82" y="110" font-family="Segoe UI, Arial, sans-serif"
        font-size="28" font-weight="700" fill="#ffffff" text-anchor="middle">
    S
  </text>

  <!-- App name -->
  <text x="82" y="175" font-family="Segoe UI, Arial, sans-serif"
        font-size="20" font-weight="600" fill="#ffffff" text-anchor="middle">
    S-AGI
  </text>
  <text x="82" y="195" font-family="Segoe UI, Arial, sans-serif"
        font-size="10" fill="rgba(255,255,255,0.6)" text-anchor="middle">
    AI Agent for
  </text>
  <text x="82" y="210" font-family="Segoe UI, Arial, sans-serif"
        font-size="10" fill="rgba(255,255,255,0.6)" text-anchor="middle">
    Spreadsheets
  </text>

  <!-- Decorative lines -->
  <line x1="30" y1="250" x2="134" y2="250" stroke="rgba(255,255,255,0.1)" stroke-width="1"/>

  <!-- Version -->
  <text x="82" y="290" font-family="Segoe UI, Arial, sans-serif"
        font-size="10" fill="rgba(255,255,255,0.4)" text-anchor="middle">
    Version 0.1.0
  </text>
</svg>`

  const svgPath = join(buildDir, "installer-sidebar.svg")
  writeFileSync(svgPath, svg)

  try {
    const sharp = await import("sharp")
    await sharp.default(Buffer.from(svg))
      .resize(164, 314)
      .png()
      .toFile(outputPath.replace('.bmp', '.png'))

    console.log(`✓ Generated NSIS sidebar: ${outputPath.replace('.bmp', '.png')}`)
    return true
  } catch (e) {
    console.log(`⚠ Could not generate NSIS sidebar. SVG saved at: ${svgPath}`)
    return false
  }
}

// Main execution
console.log("=".repeat(50))
console.log("S-AGI Installer Assets Generator")
console.log("=".repeat(50))
console.log()

await generateDMGBackground()
await generateNSISHeader()
await generateNSISSidebar()

console.log()
console.log("=".repeat(50))
console.log("Done!")
console.log("=".repeat(50))

File: apps/electron/scripts/generate-installer-assets.py
#!/usr/bin/env python3
"""
Generate installer assets using the app's background image
Creates beautiful DMG and NSIS installer graphics

Generates high-quality @2x Retina images for best display quality.
"""

from PIL import Image, ImageFilter, ImageDraw, ImageFont
import os
from pathlib import Path

# Paths
SCRIPT_DIR = Path(__file__).parent
BUILD_DIR = SCRIPT_DIR.parent / "build"
SOURCE_BG = BUILD_DIR / "background.png"

# Use @2x resolution for Retina displays
RETINA_SCALE = 2

def create_dmg_background():
    """Create DMG background with drag-to-Applications design at @2x resolution"""
    print("Creating DMG background (@2x Retina)...")

    # Base dimensions (will be scaled 2x for Retina)
    base_width, base_height = 660, 400
    width, height = base_width * RETINA_SCALE, base_height * RETINA_SCALE

    # Load and process background
    bg = Image.open(SOURCE_BG)

    # Calculate crop to cover the target size
    bg_ratio = bg.width / bg.height
    target_ratio = width / height

    if bg_ratio > target_ratio:
        # Image is wider, crop width
        new_width = int(bg.height * target_ratio)
        left = (bg.width - new_width) // 2
        bg = bg.crop((left, 0, left + new_width, bg.height))
    else:
        # Image is taller, crop height
        new_height = int(bg.width / target_ratio)
        top = (bg.height - new_height) // 2
        bg = bg.crop((0, top, bg.width, top + new_height))

    # Resize to target dimensions
    bg = bg.resize((width, height), Image.Resampling.LANCZOS)

    # Apply slight blur (scaled for resolution)
    bg = bg.filter(ImageFilter.GaussianBlur(radius=2 * RETINA_SCALE))

    # Darken slightly
    from PIL import ImageEnhance
    enhancer = ImageEnhance.Brightness(bg)
    bg = enhancer.enhance(0.85)

    # Create overlay
    overlay = Image.new('RGBA', (width, height), (0, 0, 0, 0))
    draw = ImageDraw.Draw(overlay)

    # Semi-transparent dark overlay
    draw.rectangle([0, 0, width, height], fill=(0, 0, 0, 76))  # 30% opacity

    # Scale factor for all dimensions
    s = RETINA_SCALE

    # Try to load a nice font, fallback to default (scaled for Retina)
    try:
        title_font = ImageFont.truetype("/System/Library/Fonts/Helvetica.ttc", 32 * s)
        subtitle_font = ImageFont.truetype("/System/Library/Fonts/Helvetica.ttc", 14 * s)
        label_font = ImageFont.truetype("/System/Library/Fonts/Helvetica.ttc", 13 * s)
        instruction_font = ImageFont.truetype("/System/Library/Fonts/Helvetica.ttc", 15 * s)
        version_font = ImageFont.truetype("/System/Library/Fonts/Helvetica.ttc", 12 * s)
    except:
        title_font = ImageFont.load_default()
        subtitle_font = title_font
        label_font = title_font
        instruction_font = title_font
        version_font = title_font

    # Title
    draw.text((width//2, 50 * s), "S-AGI", font=title_font, fill=(255, 255, 255, 255), anchor="mm")
    draw.text((width//2, 78 * s), "AI Agent for Spreadsheets", font=subtitle_font, fill=(255, 255, 255, 200), anchor="mm")

    # Drop zone boxes (glassmorphism style)
    # App icon zone (left)
    box_size = 110 * s
    app_x, app_y = 180 * s, 175 * s
    draw.rounded_rectangle(
        [app_x - box_size//2, app_y - box_size//2, app_x + box_size//2, app_y + box_size//2],
        radius=24 * s,
        fill=(255, 255, 255, 25),
        outline=(255, 255, 255, 76),
        width=2 * s
    )

    # Applications zone (right)
    apps_x, apps_y = 480 * s, 175 * s
    draw.rounded_rectangle(
        [apps_x - box_size//2, apps_y - box_size//2, apps_x + box_size//2, apps_y + box_size//2],
        radius=24 * s,
        fill=(255, 255, 255, 25),
        outline=(255, 255, 255, 76),
        width=2 * s
    )

    # Arrow between boxes
    arrow_y = 175 * s
    arrow_start = app_x + box_size//2 + 20 * s
    arrow_end = apps_x - box_size//2 - 20 * s

    # Dashed line effect (draw segments)
    dash_length = 12 * s
    gap_length = 6 * s
    x = arrow_start
    while x < arrow_end - 20 * s:
        draw.line([(x, arrow_y), (min(x + dash_length, arrow_end - 20 * s), arrow_y)],
                  fill=(255, 255, 255, 153), width=3 * s)
        x += dash_length + gap_length

    # Arrow head
    arrow_head_x = arrow_end
    draw.polygon([
        (arrow_head_x, arrow_y),
        (arrow_head_x - 20 * s, arrow_y - 14 * s),
        (arrow_head_x - 20 * s, arrow_y + 14 * s)
    ], fill=(255, 255, 255, 178))

    # Labels
    draw.text((180 * s, 270 * s), "S-AGI.app", font=label_font, fill=(255, 255, 255, 230), anchor="mm")
    draw.text((480 * s, 270 * s), "Applications", font=label_font, fill=(255, 255, 255, 230), anchor="mm")

    # Instructions
    draw.text((width//2, 330 * s), "Drag S-AGI to Applications to install",
              font=instruction_font, fill=(255, 255, 255, 216), anchor="mm")

    # Version badge
    badge_width = 80 * s
    badge_height = 26 * s
    badge_x = width//2 - badge_width//2
    badge_y = 350 * s
    draw.rounded_rectangle(
        [badge_x, badge_y, badge_x + badge_width, badge_y + badge_height],
        radius=13 * s,
        fill=(255, 255, 255, 38)
    )
    draw.text((width//2, badge_y + badge_height//2), "v0.1.0", font=version_font, fill=(255, 255, 255, 178), anchor="mm")

    # Composite
    bg = bg.convert('RGBA')
    result = Image.alpha_composite(bg, overlay)
    result = result.convert('RGB')

    # Save at maximum quality (compress_level=0 = no compression)
    result.save(BUILD_DIR / "dmg-background.png", "PNG", compress_level=0)
    print(f"✓ DMG background created ({width}x{height} @2x Retina)")

def create_nsis_sidebar():
    """Create NSIS installer sidebar at high resolution"""
    print("Creating NSIS sidebar (high quality)...")

    # NSIS sidebar is 164x314 - use 2x for better quality on high-DPI displays
    base_width, base_height = 164, 314
    width, height = base_width * RETINA_SCALE, base_height * RETINA_SCALE

    # Load and process background
    bg = Image.open(SOURCE_BG)

    # Crop and resize
    bg_ratio = bg.width / bg.height
    target_ratio = width / height

    if bg_ratio > target_ratio:
        new_width = int(bg.height * target_ratio)
        left = (bg.width - new_width) // 2
        bg = bg.crop((left, 0, left + new_width, bg.height))
    else:
        new_height = int(bg.width / target_ratio)
        top = (bg.height - new_height) // 2
        bg = bg.crop((0, top, bg.width, top + new_height))

    bg = bg.resize((width, height), Image.Resampling.LANCZOS)
    bg = bg.filter(ImageFilter.GaussianBlur(radius=3 * RETINA_SCALE))

    from PIL import ImageEnhance
    enhancer = ImageEnhance.Brightness(bg)
    bg = enhancer.enhance(0.7)

    # Scale factor
    s = RETINA_SCALE

    # Create overlay
    overlay = Image.new('RGBA', (width, height), (0, 0, 0, 0))
    draw = ImageDraw.Draw(overlay)

    # Gradient overlay (darker at bottom)
    for y in range(height):
        alpha = int(100 + (y / height) * 50)  # 40% to 60%
        draw.line([(0, y), (width, y)], fill=(0, 0, 0, alpha))

    # Try to load fonts (scaled)
    try:
        logo_font = ImageFont.truetype("/System/Library/Fonts/Helvetica.ttc", 36 * s)
        name_font = ImageFont.truetype("/System/Library/Fonts/Helvetica.ttc", 22 * s)
        desc_font = ImageFont.truetype("/System/Library/Fonts/Helvetica.ttc", 11 * s)
        version_font = ImageFont.truetype("/System/Library/Fonts/Helvetica.ttc", 10 * s)
    except:
        logo_font = ImageFont.load_default()
        name_font = logo_font
        desc_font = logo_font
        version_font = logo_font

    # Logo circle
    cx, cy = width // 2, 100 * s
    r = 45 * s
    draw.ellipse([cx - r, cy - r, cx + r, cy + r], fill=(255, 255, 255, 38), outline=(255, 255, 255, 76), width=2 * s)
    draw.text((cx, cy + 5 * s), "S", font=logo_font, fill=(255, 255, 255, 255), anchor="mm")

    # App name
    draw.text((width//2, 180 * s), "S-AGI", font=name_font, fill=(255, 255, 255, 255), anchor="mm")
    draw.text((width//2, 202 * s), "AI Agent for", font=desc_font, fill=(255, 255, 255, 178), anchor="mm")
    draw.text((width//2, 218 * s), "Spreadsheets", font=desc_font, fill=(255, 255, 255, 178), anchor="mm")

    # Decorative line
    draw.line([(30 * s, 250 * s), (134 * s, 250 * s)], fill=(255, 255, 255, 51), width=1 * s)

    # Version
    draw.text((width//2, 290 * s), "Version 0.1.0", font=version_font, fill=(255, 255, 255, 127), anchor="mm")

    # Composite
    bg = bg.convert('RGBA')
    result = Image.alpha_composite(bg, overlay)

    # Resize back to target size for Windows (which doesn't need @2x)
    result = result.resize((base_width, base_height), Image.Resampling.LANCZOS)
    result = result.convert('RGB')

    result.save(BUILD_DIR / "installer-sidebar.png", "PNG", compress_level=0)
    print(f"✓ NSIS sidebar created ({base_width}x{base_height}, rendered @{RETINA_SCALE}x)")

def create_nsis_header():
    """Create NSIS installer header at high resolution"""
    print("Creating NSIS header (high quality)...")

    # NSIS header is 150x57 - render at 2x for quality
    base_width, base_height = 150, 57
    width, height = base_width * RETINA_SCALE, base_height * RETINA_SCALE
    s = RETINA_SCALE

    # Load and process background
    bg = Image.open(SOURCE_BG)

    # Crop top portion and resize
    bg_ratio = bg.width / bg.height
    target_ratio = width / height

    if bg_ratio > target_ratio:
        new_width = int(bg.height * target_ratio)
        left = (bg.width - new_width) // 2
        bg = bg.crop((left, 0, left + new_width, bg.height))
    else:
        new_height = int(bg.width / target_ratio)
        bg = bg.crop((0, 0, bg.width, new_height))

    bg = bg.resize((width, height), Image.Resampling.LANCZOS)
    bg = bg.filter(ImageFilter.GaussianBlur(radius=2 * s))

    from PIL import ImageEnhance
    enhancer = ImageEnhance.Brightness(bg)
    bg = enhancer.enhance(0.75)

    # Create overlay
    overlay = Image.new('RGBA', (width, height), (0, 0, 0, 76))
    draw = ImageDraw.Draw(overlay)

    try:
        title_font = ImageFont.truetype("/System/Library/Fonts/Helvetica.ttc", 20 * s)
    except:
        title_font = ImageFont.load_default()

    draw.text((width//2, height//2 + 2 * s), "S-AGI", font=title_font, fill=(255, 255, 255, 255), anchor="mm")

    # Composite
    bg = bg.convert('RGBA')
    result = Image.alpha_composite(bg, overlay)

    # Resize back to target size for Windows
    result = result.resize((base_width, base_height), Image.Resampling.LANCZOS)
    result = result.convert('RGB')

    result.save(BUILD_DIR / "installer-header.png", "PNG", compress_level=0)
    print(f"✓ NSIS header created ({base_width}x{base_height}, rendered @{RETINA_SCALE}x)")

if __name__ == "__main__":
    print("=" * 50)
    print("S-AGI Installer Assets Generator")
    print("Using app background image")
    print("=" * 50)
    print()

    create_dmg_background()
    create_nsis_sidebar()
    create_nsis_header()

    print()
    print("=" * 50)
    print("All assets generated successfully!")
    print("=" * 50)
